Carousel: improve gallery parsing

jeherve opened the pull request:
Fixes #13531
Fixes #13428 
Fixes #11191 
Fixes #17090 

#### Changes proposed in this Pull Request:

**This is still a work in progress. I'd like to add Unit Tests to cover each possible use of Carousel, as detailed in the testing instructions below.**

Parsing post content to insert carousel metadata is a bit fragile at the moment. Depending on the format of your images, how they are inserted, and what metadata is available for each image, you may run into several issues, highlighted in the issues above.

This PR introduces a different approach to add the Carousel metadata to the classic galleries (it hooks into the `do_shortcode_tag` filter instead of `gallery_style`).
It also takes on @mdawaffe's approach from #13547 to improve our use of `DOMDocument`.
It also fixes PHPCS errors to get things to be more readable as I was going through that file.

#### Jetpack product discussion

* N/A

#### Does this pull request change what data or activity we track or use?

* No

#### Testing instructions:

* In new posts, insert a variety of blocks:
	- A classic block with a gallery
	- A classic block with a tiled gallery
	- A classic block with a single image
	- An image block
	- A gallery block
	- A Tiled Gallery block
	- A column block with some text
	- A column block with a gallery block in it.
* Publish all those posts
* When moving your mouse over each image, make sure the cursor only becomes a pointer when the element can be expanded to a Carousel modal.
* Carousel should work well for all those blocks.
* You should not see any JavaScript error.

mdawaffe commented:
I haven't checked, but I think the problems from https://github.com/Automattic/jetpack/pull/13547#issuecomment-535281664 are the same in this PR as they were in the other.

mdawaffe commented on modules/carousel/jetpack-carousel.php:
```diff
@@ -547,16 +682,28 @@ class_exists( 'Jetpack_AMP_Support' )
 				// Let's grab all containers from the HTML.
 				$dom_doc = new DOMDocument();
 
+				$dom_doc->encoding = $charset;
```
I don't think this line is necessary. I added it to see what it might change, but I don't think it has any effect (the `<meta>` stuff below is sufficient). Testing may prove me wrong :)

jeherve committed:
Carousel: refactor to improve gallery parsing.

- Remove in_gallery; it hasn't been used since #5469

- add comments, docblocks, and fix various PHPCS errors.

- use the_content to inspect all posts for galleries

In the past we also relied on the gallery_style filter, which while it works, only relies on the gallery's opening div tag. Since we're already examining the post content for gallery blocks, let's extend that to check for old-style galleries as well.

- stop saveHTML from adding unneeded doctype and html tags

- improve DOM Parsing
    - Wrap input to `DOMDocument->loadHTML()` in fake root tag.
    - Specify blog's charset in input to `DOMDocument->loadHTML()`.

- filter final output of gallery shortcode to add metadata

This replaces the 2 other approaches we had tried:
- Before this PR, we relied on the gallery_style filter. This was not ideal as the filter contains only the opening part of the div tag.
- Earlier in this PR I tried to rely on the_content, but at that point the gallery shortcode is still just a shortcode, so we can't add metadata yet.

- Add first set of unit tests

jeherve committed:
Add changelog

zinigor commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.php:
```diff
@@ -656,10 +779,75 @@ class_exists( 'Jetpack_AMP_Support' )
-				$html = str_replace( '<ul class="blocks-gallery-grid', '<ul ' . esc_attr( $data_key ) . "='" . wp_json_encode( $data_values ) . "' class=\"blocks-gallery-grid", $html );
-				$html = str_replace( '<figure class="wp-block-gallery blocks-gallery-grid', '<figure ' . esc_attr( $data_key ) . "='" . wp_json_encode( $data_values ) . "' class=\"wp-block-gallery  blocks-gallery-grid", $html );
+				// Do not go any further if DOMDocument is disabled on the server.
+				if ( ! class_exists( 'DOMDocument' ) ) {
```
Do you think it makes sense to return before `foreach`? Maybe even before anything else is checked to save us an autoloader roundtrip.

zinigor commented:
Great stuff, sorry, couldn't help myself and did a drive-by review. My only other comment about the code is that I like your intention to add more stuff to test with. It would be great to test with some gibberish too, since the HTML can be put through any sort of filter before it gets to the carousel. So that if DOMDocument freaks out, we can fail gracefully.
